### PR TITLE
Collector refactor

### PR DESCRIPTION
**Pull request by LouisYRYJ**
This PR introduces the new Collector class that will make it easy to implement custom hooks, e.g. for EK-FAC and generally allow for more flexible custom methods

### CODE REVIEW

File: bergson/collector/logger.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# Set up logging
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(filename)s:%(funcName)s:%(lineno)d:%(levelname)s:  %(message)s",
+    stream=sys.stdout,  # Ensure output goes to stdout
+    force=True,  # Override any existing configuration
+)
+
+
+class RankFilter(logging.Filter):
```
Comment (luciaquirke) [inline bergson/collector/logger.py]: Cool!

File: bergson/collector/collector.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+    """
+    Set of module names to attach hooks to. Should consist only of nn.Linear modules.
+    If None, hooks are attached to all Linear layers in the model.
+    """
+
+    processor: GradientProcessor = field(default_factory=GradientProcessor)
+    """Configuration for processing and compressing gradients."""
+
+    # TODO: Understand this
+    indices: list[int] = field(default_factory=list)
```
Comment (luciaquirke) [inline bergson/collector/collector.py]: I think this can be removed now

File: bergson/collection.py [hunk 3]
```diff
@@ -24,247 +19,32 @@ def collect_gradients(
     *,
     batches: list[list[int]] | None = None,
     target_modules: set[str] | None = None,
-    attention_cfgs: dict[str, AttentionConfig] | None = None,
+    attention_cfgs: dict[str, AttentionConfig] = {},
```
[the remark this answers is not in the document]
Comment (LouisYRYJ) [reply bergson/collection.py]: good point will change!

File: bergson/collector/collector.py [hunk 4]
```diff
@@ -0,0 +1,734 @@
...
+from bergson.utils import assert_type, create_projection_matrix
+
+
+@dataclass
+class HookCollectorBase(ContextDecorator, ABC):
+    """
+    Abstract base class for collectors that attach forward and backward hooks to model
+    layers.
+
+    Automatically discovers nn.Linear layers in the model, registers hooks during
```
Comment (luciaquirke) [inline bergson/collector/collector.py]: nn.Linear layers -> supported modules

File: bergson/collector/collector.py [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+        - backward_hook(): Process gradients during backward pass
+    """
+
+    model: nn.Module
+
+    cfg: IndexConfig
+
+    target_modules: set[str] | None = None
+    """
+    Set of module names to attach hooks to. Should consist only of nn.Linear modules.
```
Comment (luciaquirke) [inline bergson/collector/collector.py]: Fix

File: bergson/collector/collector.py [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+
+    mod_grads: dict = field(default_factory=dict)
+
+    precondition: bool = False
+
+    device: torch.device | str
+
+    dtype: torch.dtype
+
+    def setup(self) -> None:
```
Comment (luciaquirke) [inline bergson/collector/collector.py]: Is this removable btw

File: bergson/collector/collector.py [hunk 7]
```diff
@@ -0,0 +1,779 @@
...
+    def __post_init__(
+        self,
+    ):
+        self.rank = dist.get_rank() if dist.is_initialized() else 0
+        self.world_size = dist.get_world_size() if dist.is_initialized() else 1
+
+        self._fwd_hooks: list[RemovableHandle] = []
+        self._bwd_hooks: list[RemovableHandle] = []
+
+        # Discover target Linear modules using the static method
```
Comment (luciaquirke) [inline bergson/collector/collector.py]:
```suggestion
        # Discover target modules using the static method
```

File: bergson/collector/collector.py [hunk 8]
```diff
@@ -0,0 +1,779 @@
...
+    I=input dimension.
+
+    Subclasses must implement:
+        - setup(): Initialize state (buffers, dicts, etc.)
+        - teardown(): Clean up and save results
+        - forward_hook(): Process activations during forward pass
+        - backward_hook(): Process gradients during backward pass
+    """
+
+    model: nn.Module
```
Comment (luciaquirke) [inline bergson/collector/collector.py]:
```suggestion
    model: nn.Module
    """Model that forward and backward hooks will be attached to."""
```

File: bergson/collector/collector.py [hunk 9]
```diff
@@ -0,0 +1,779 @@
...
+        """
+        pass
+
+
+@dataclass(kw_only=True)
+class GradientCollector(HookCollectorBase):
+    """
+    Collects gradients for each Linear layer in the model.
+
+    Stores gradients in a dictionary mapping module names to gradient tensors.
```
Comment (luciaquirke) [inline bergson/collector/collector.py]: TODO 
For each step/gradient accumulation step/whatever, store gradients in...

**Comment by luciaquirke** [comment]
Reminder docstrings
